### DISCUSSION TITLE
Maintaining history of changes in settings

### DISCUSSION THREAD

**Discussion by wntechs** [spatie/laravel-settings#236] [Ideas]
This package is a good peice of work. For most use cases it's sufficient. However I strongly feel that the settings change over time and at some obvious point in time, the historical changes get lost with updated values.

Here is an exmplae usage-

Suppose there is a system for rewarding its customers a certain cashback points based on the puchases they have in last month.
The configuration for how much percentage of actual purchase will be rewarded as cashback is kept in settings table.
This fine as long as the percentage doesn't change. But as soon as we change the percentage, we lost track of what was the rewarded percentage for old purchases.
It makes me think that if there would have been a version number in the settings we could easily use that number while rewarding cashbacks and anytime we need to figure out the configuration for past rewards we can easily link them up with the version number.

Technically the settings classes whould accept a version number and populate the values based on this. 

I hope this makes sense.
Thanks

**Comment by rubenvanassche** [comment]
Take a look at event sourcing, that's probably a good solution to this problem. Use data objects to store the percentages at a point in time which are copied from the settings. That keeps your settings dynamic and you can also have different versions of their values.

**Reply by wntechs** [reply]
Thanks @rubenvanassche for the workaround to the concern. I am kind of following the same thing currently by taking a snapshot of the settings table and saving it into another table to log the history. This is fine if we are to investigate the setting history manually. But it's a bit hard to map the historical data into setting class. The setting classes will always pickup the latest data from settings table and not from the history table. Is there any cleaner way to reconstruct the setting classes based on the historical data?
Thanks again for checking on this.

**Reply by rubenvanassche** [reply]
Settings class is filled up by a settings repository, you could write your own one and point it to another table.

**Reply by wntechs** [reply]
> Settings class is filled up by a settings repository, you could write your own one and point it to another table.

Ok thank you for the right directions. I will check to create settings repository and see if it works.
